File: Design-Patterns-Using-Swift/Builder/BuilderPattern.swift
// Define the Product class with two parts
class Product {
    private var partA: String
    private var partB: String

    init(A: String, B: String) {
        self.partA = A
        self.partB = B
    }

    func setPartA(_ A: String) {
        self.partA = A
    }

    func setPartB(_ B: String) {
        self.partB = B
    }

    var description: String {
        return "Product : (\(partA), \(partB))"
    }
}

// Define an abstract class called Builder
class Builder {
    var product: Product

    init() {
        self.product = Product(A: "A default", B: "B default")
    }

    func setPartA(_ A: String) -> Builder {
        fatalError("Abstract method. Subclasses must implement this method.")
    }

    func setPartB(_ B: String) -> Builder {
        fatalError("Abstract method. Subclasses must implement this method.")
    }

    func getProduct() -> Product {
        let temp = self.product
        self.product = Product(A: "A default", B: "B default") // assign new product.
        return temp
    }
}

// Define a ConcreteBuilder class that extends Builder
class ConcreteBuilder: Builder {
    override func setPartA(_ A: String) -> Builder {
        self.product.setPartA(A)
        return self
    }

    override func setPartB(_ B: String) -> Builder {
        self.product.setPartB(B)
        return self
    }
}

// Define a Director class that takes a builder object as a parameter
class Director {
    private var builder: Builder

    init(builder: Builder) {
        self.builder = builder
    }

    func construct() -> Product {
        return self.builder.setPartA("A1").setPartB("B1").getProduct()
    }

    func construct2() -> Product {
        _ = self.builder.setPartA("A2").setPartB("B2")
        return self.builder.getProduct()
    }

    func construct3() -> Product {
        return self.builder.setPartA("A3").getProduct()
    }
}

// Client code
let builder = ConcreteBuilder()
let director = Director(builder: builder)

let product = director.construct()
print(product.description)

let product2 = director.construct2()
print(product2.description)

let product3 = director.construct3()
print(product3.description)

/*
Product : (A1, B1)
Product : (A2, B2)
Product : (A3, B default)
*/

File: Design-Patterns-Using-JavaScript/Builder/BuilderPattern.js
// Define the Product class with two parts
class Product {
    constructor(A = "A default", B = "B default") {
        this.partA = A;
        this.partB = B;
    }

    setPartA(A) {
        this.partA = A;
    }

    setPartB(B) {
        this.partB = B;
    }

    toString() {
        return `Product : (${this.partA}, ${this.partB})`;
    }
}

// Define an abstract class called Builder
class Builder {
    constructor() {
        this.product = new Product();
    }

    setPartA(A) {
        throw new Error("Abstract method: setPartA");
    }

    setPartB(B) {
        throw new Error("Abstract method: setPartB");
    }

    getProduct() {
        const temp = this.product;
        this.product = new Product(); // assign new product.
        return temp;
    }
}

// Define a ConcreteBuilder class that extends Builder
class ConcreteBuilder extends Builder {
    setPartA(A) {
        this.product.setPartA(A);
        return this;
    }

    setPartB(B) {
        this.product.setPartB(B);
        return this;
    }
}

// Define a Director class that takes a builder object as a parameter
class Director {
    constructor(builder) {
        this.builder = builder;
    }

    construct() {
        return this.builder.setPartA("A1").setPartB("B1").getProduct();
    }

    construct2() {
        this.builder.setPartA("A2");
        this.builder.setPartB("B2");
        return this.builder.getProduct();
    }

    construct3() {
        return this.builder.setPartA("A3").getProduct();
    }
}

// Client code
const builder = new ConcreteBuilder();
const director = new Director(builder);

const product = director.construct();
console.log(product.toString());

const product2 = director.construct2();
console.log(product2.toString());

const product3 = director.construct3();
console.log(product3.toString());

/*
Product : (A1, B1)
Product : (A2, B2)
Product : (A3, B default)
*/